File: src/storage.ts
import { readFileSync, writeFileSync } from "fs";
import * as fsPath from "path";
import { DATA_FILE_PATH } from "./constants";
import { BranchData, BranchDataCollection, Result, ok, err } from "./types";

// --- adapters

/**
 * Reads jump data from disk, cleans up stale branches, saves the cleaned data
 * back to disk, and returns the reconciled branch list.
 */
export function getAndCleanBranchData(
  rawGitBranches: string[],
  gitRepoFolder: string,
): Result<BranchData[]> {
  const res1 = readBranchesJumpData(gitRepoFolder);
  if (res1.tag === "err") return res1;

  const cleanJumpData = filterStaleJumpData(rawGitBranches, res1.value);
  const reconciledBranches = reconcileBranches(rawGitBranches, cleanJumpData);

  const res2 = saveBranchesJumpData(gitRepoFolder, cleanJumpData);
  if (res2.tag === "err") return res2;

  return ok(reconciledBranches);
}

/**
 * Reads the historical data from disk, updates the timestamp in memory,
 * and immediately flushes the changes back to `.jump/data.json`.
 */
export function updateBranchLastSwitch(
  name: string,
  lastSwitch: number,
  gitRepoFolder: string,
): Result<void> {
  const res1 = readBranchesJumpData(gitRepoFolder);
  if (res1.tag === "err") return res1;

  const updatedData = setBranchTimestamp(res1.value, name, lastSwitch);

  return saveBranchesJumpData(gitRepoFolder, updatedData);
}

/**
 * Reads the current data from disk, applies the pure rename transformation,
 * and safely writes the updated history back to `.jump/data.json`.
 */
export function renameJumpDataBranch(
  currentName: string,
  newName: string,
  gitRepoFolder: string,
): Result<void> {
  const res1 = readBranchesJumpData(gitRepoFolder);
  if (res1.tag === "err") return res1;

  const updatedData = renameBranch(res1.value, currentName, newName);

  return saveBranchesJumpData(gitRepoFolder, updatedData);
}

/**
 * Reads the current data from disk, filters out the specified branches,
 * and writes the clean data back to `.jump/data.json`.
 */
export function deleteJumpDataBranch(
  branchNames: string[],
  gitRepoFolder: string,
): Result<void> {
  const res1 = readBranchesJumpData(gitRepoFolder);
  if (res1.tag === "err") return res1;

  const updatedData = deleteBranches(res1.value, branchNames);

  return saveBranchesJumpData(gitRepoFolder, updatedData);
}

// --- actual file I/O

function readBranchesJumpData(
  gitRepoFolder: string,
): Result<BranchDataCollection> {
  try {
    return ok(
      JSON.parse(
        readFileSync(fsPath.join(gitRepoFolder, DATA_FILE_PATH)).toString(),
      ),
    );
  } catch (e) {
    return err(
      new Error(
        `JSON in "${DATA_FILE_PATH}" is not valid, could not parse it.`,
      ),
    );
  }
}

function saveBranchesJumpData(
  gitRepoFolder: string,
  jumpData: BranchDataCollection,
): Result<void> {
  try {
    writeFileSync(
      fsPath.join(gitRepoFolder, DATA_FILE_PATH),
      JSON.stringify(jumpData, null, 2),
    );
    return ok(undefined);
  } catch (e) {
    return err(new Error(`Could not write data into "${DATA_FILE_PATH}".`));
  }
}

// --- pure utils

function reconcileBranches(
  rawGitBranches: string[],
  jumpData: BranchDataCollection,
): BranchData[] {
  return rawGitBranches.map((branch) => ({
    name: branch,
    lastSwitch: jumpData[branch]?.lastSwitch ?? 0,
  }));
}

function filterStaleJumpData(
  rawGitBranches: string[],
  jumpData: BranchDataCollection,
): BranchDataCollection {
  return Object.fromEntries(
    Object.entries(jumpData).filter(([b, _]) => rawGitBranches.includes(b)),
  );
}

function setBranchTimestamp(
  jumpData: BranchDataCollection,
  name: string,
  lastSwitch: number,
): BranchDataCollection {
  return {
    ...jumpData,
    [name]: { name, lastSwitch },
  };
}

function renameBranch(
  jumpData: BranchDataCollection,
  currentName: string,
  newName: string,
): BranchDataCollection {
  const currentJumpData = jumpData[currentName];
  if (!currentJumpData) return jumpData;

  const { [currentName]: _removed, ...rest } = jumpData;

  return {
    ...rest,
    [newName]: { ...currentJumpData, name: newName },
  };
}

function deleteBranches(
  jumpData: BranchDataCollection,
  branchesToDelete: string[],
): BranchDataCollection {
  return Object.fromEntries(
    Object.entries(jumpData).filter(([b, _]) => !branchesToDelete.includes(b)),
  );
}

File: git-jump/src/storage.rs
use std::{
    collections::{HashMap, HashSet},
    fs::{self, File},
    io::{BufReader, BufWriter},
    path::Path,
};

use crate::types::Branch;
use anyhow::{Context, Result};
use serde::Deserialize;

pub fn get_and_clean_branches(data_file: &Path, branch_names: &[&str]) -> Result<Vec<Branch>> {
    let mut jump_data: BranchCollection = load_jump_data(&data_file)?;

    keep_branches(&mut jump_data, branch_names);
    let reconciled_branches = reconcile_branches(&jump_data, branch_names);

    save_branches_jump_data(&data_file, &jump_data)?;

    Ok(reconciled_branches)
}

type BranchCollection = HashMap<String, u64>;

/// Reads the historical data from disk, updates the timestamp in memory,
/// and immediately flushes the changes back to `.jump/data.json`.
pub fn update_branch_last_switch(data_file: &Path, name: &str, last_switch: u64) -> Result<()> {
    let mut jump_data: BranchCollection = load_jump_data(&data_file)?;

    set_branch_timestamp(&mut jump_data, name, last_switch);

    save_branches_jump_data(&data_file, &jump_data)
}

/// Reads the current data from disk, applies the pure rename transformation,
/// and safely writes the updated history back to `.jump/data.json`.
pub fn rename_jump_data_branch(data_file: &Path, current_name: &str, new_name: &str) -> Result<()> {
    let mut jump_data: BranchCollection = load_jump_data(&data_file)?;

    rename_branch(&mut jump_data, current_name, new_name);

    save_branches_jump_data(&data_file, &jump_data)
}

/// Reads the current data from disk, filters out the specified branches,
/// and writes the clean data back to `.jump/data.json`.
pub fn delete_jump_data_branch(data_file: &Path, branch_names: &[&str]) -> Result<()> {
    let mut jump_data: BranchCollection = load_jump_data(&data_file)?;

    delete_branches(&mut jump_data, branch_names);

    save_branches_jump_data(&data_file, &jump_data)
}

// --- actual file I/O

#[derive(Deserialize)]
#[serde(untagged)]
enum OnDisk {
    V2(HashMap<String, u64>),
    V1(HashMap<String, BranchEntry>),
}

#[derive(Deserialize)]
struct BranchEntry {
    #[serde(rename = "lastSwitch")]
    last_switch: u64,
}

fn parse_from_disk(data: OnDisk) -> BranchCollection {
    match data {
        OnDisk::V2(hash_map) => hash_map,
        OnDisk::V1(hash_map) => hash_map
            .into_iter()
            .map(|(k, v)| (k, v.last_switch))
            .collect::<BranchCollection>(),
    }
}

/// Loads jump data from disk, normalising the on-disk format if needed.
///
/// Supports both:
/// - the legacy V1 format (`{ "branch": { "name": <string>, "lastSwitch": <timestamp> } }`)
/// - the current V2 format (`{ "branch": <timestamp> }`)
///
/// When a V1 file is encountered, the original is copied to `<data_file>.v1.bak`
/// before the caller writes it back in V2 form.
fn load_jump_data(data_file: &Path) -> Result<BranchCollection> {
    let file = File::open(data_file)?;
    let reader = BufReader::new(file);

    let branches: OnDisk =
        serde_json::from_reader(reader).context("Failed to deserialize jump data")?;

    match branches {
        OnDisk::V2(_) => {}
        OnDisk::V1(_) => {
            // This is the aforementioned side-effect
            create_backup(data_file)?;
        }
    };

    Ok(parse_from_disk(branches))
}

fn create_backup(data_file: &Path) -> Result<()> {
    let backup_file = data_file.with_extension(".json.v1.bak");
    fs::copy(data_file, backup_file).context("Failed to back up old jump data file")?;
    Ok(())
}

fn save_branches_jump_data(data_file: &Path, jump_data: &BranchCollection) -> Result<()> {
    let file = File::open(data_file)?;
    let writer = BufWriter::new(file);
    serde_json::to_writer(writer, &jump_data).context("Failed to write serialized branch data")?;
    Ok(())
}

// pure utils

fn reconcile_branches(jump_data: &BranchCollection, branch_names: &[&str]) -> Vec<Branch> {
    branch_names
        .iter()
        .map(|&branch_name| Branch {
            name: branch_name.to_owned(),
            last_switch: jump_data.get(branch_name).copied().unwrap_or(0u64),
        })
        .collect::<Vec<Branch>>()
}

fn set_branch_timestamp(jump_data: &mut BranchCollection, name: &str, last_switch: u64) {
    jump_data.insert(name.to_owned(), last_switch);
}

fn rename_branch(jump_data: &mut BranchCollection, current_name: &str, new_name: &str) {
    if let Some(last_switch) = jump_data.remove(current_name) {
        jump_data.insert(new_name.to_owned(), last_switch);
    };
}

enum FilterMode {
    Keep,
    Delete,
}

fn filter_jump_data(jump_data: &mut BranchCollection, branch_names: &[&str], mode: FilterMode) {
    let set: HashSet<&str> = branch_names.iter().copied().collect();
    jump_data.retain(|k, _| {
        let in_set = set.contains(k.as_str());
        match mode {
            FilterMode::Keep => in_set,
            FilterMode::Delete => !in_set,
        }
    });
}

fn keep_branches(jump_data: &mut BranchCollection, branch_names: &[&str]) {
    filter_jump_data(jump_data, branch_names, FilterMode::Keep);
}

fn delete_branches(jump_data: &mut BranchCollection, branch_names: &[&str]) {
    filter_jump_data(jump_data, branch_names, FilterMode::Delete);
}
